File: swift-package/SkylogKit/Sources/SkylogKit/writers/InMemoryLogWriter.swift
import Foundation
import Combine

/// A `LogWriter` that retains the most recent `capacity` entries in a bounded ring buffer.
///
/// Mirrors `dev.viethung.skylog.writers.InMemoryLogWriter` from `:skylog-core`, substituting:
/// - `StateFlow<List<LogEntry>>` → `@Published var entries: [LogEntry]` (Combine / `ObservableObject`).
/// - `MutableStateFlow.update {}` CAS → `NSLock`-guarded serial mutation.
///
/// **Thread safety:** writes and reads of `_entries` are serialized through `lock`.
/// `@Published` fires on the thread that mutates `entries`; callers driving SwiftUI should
/// ensure mutations happen on the main thread, or use `.receive(on: RunLoop.main)` on the publisher.
///
/// **Capacity:** `capacity` must be > 0. The writer silently drops the oldest entry when full.
public final class InMemoryLogWriter: LogWriter, ObservableObject {

    /// The most recent log entries, newest last. Observe via `$entries` or `objectWillChange`.
    @Published public private(set) var entries: [LogEntry] = []

    private let capacity: Int
    private let lock = NSLock()

    /// Create a writer with a bounded ring buffer.
    ///
    /// - Parameter capacity: Maximum number of entries to retain. Must be > 0.
    public init(capacity: Int = 1000) {
        precondition(capacity > 0, "InMemoryLogWriter capacity must be > 0, was \(capacity)")
        self.capacity = capacity
    }

    // MARK: - LogWriter

    public func log(_ entry: LogEntry) {
        lock.lock()
        defer { lock.unlock() }
        if entries.count < capacity {
            entries.append(entry)
        } else {
            entries.removeFirst()
            entries.append(entry)
        }
    }

    // MARK: - Housekeeping

    /// Remove all retained entries.
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        entries = []
    }

    /// Restore a previously captured snapshot — supports the console's 5-second undo flow.
    ///
    /// The snapshot is clamped to `capacity` (newest `capacity` entries) to preserve the
    /// ring-buffer invariant. Mirrors `InMemoryLogWriter.restore()` in `:skylog-core` §4.7.
    public func restore(snapshot: [LogEntry]) {
        lock.lock()
        defer { lock.unlock() }
        entries = Array(snapshot.suffix(capacity))
    }
}

File: skylog-core/src/commonMain/kotlin/dev/viethung/skylog/writers/InMemoryLogWriter.kt
package dev.viethung.skylog.writers

import dev.viethung.skylog.LogEntry
import dev.viethung.skylog.LogWriter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update

class InMemoryLogWriter(private val capacity: Int = 1000) : LogWriter() {
    init {
        require(capacity > 0) { "capacity must be > 0, was $capacity" }
    }

    private val _entries = MutableStateFlow<List<LogEntry>>(emptyList())
    val entries: StateFlow<List<LogEntry>> = _entries

    override fun log(entry: LogEntry) {
        // MutableStateFlow.update uses compareAndSet internally — retries on
        // concurrent mutation, no thread can lose an entry to a race. This is
        // the chosen serialization primitive (Mega Plan Review §2 finding
        // F-2.1: original Mutex-based plan was suspending and incompatible
        // with non-suspend log(); update() gives the same mutual-exclusion
        // semantics without going suspending).
        _entries.update { current ->
            if (current.size < capacity) current + entry
            else current.drop(1) + entry
        }
    }

    fun clear() {
        _entries.value = emptyList()
    }

    /**
     * Restore a previously captured snapshot — supports `LogConsoleScreen` undo
     * (PRD §8 + Eng Review patch P10a). Snapshot is clamped to [capacity] to
     * preserve the ring-buffer invariant. Not exposed via [LogWriter] — internal
     * to [InMemoryLogWriter].
     */
    fun restore(snapshot: List<LogEntry>) {
        _entries.value = snapshot.takeLast(capacity)
    }
}
